src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+        self_attention,  // Implies K,V lengths equaual and equal to sequence_length
+        cross_attention, // Relevant as K/V may have different lengths in this case
+    };
+
+    // For items explicitly parsed from attributes
+    struct attention_attr
+    {
+        bool do_rotary = false; // Rotary encode input prior to projection with weight matrices
+        bool past_present_share_buffer =
+            false; // Related to past input shares buffer with present output
+        bool unidirectional =
+            false; // Mask is lower triangular ie) only pay attention to prev words
+        std::size_t num_heads            = 1; // Required by inputs
+        std::size_t rotary_embedding_dim = 0; // Gets set to head_size when not set
+        std::vector<std::size_t> qkv_hidden_sizes{
+            0, 0, 0};      // Sets hidden sizes if not set defiend by input
+        float scale = 0.0; // Default 1/sqrt(query_size) (query_size also known as head_size)
+        float mask_filter_val = -10000.0f; // Default value used for masking our input encoding
+    };
+
+    // Values inferred from input vectors or attributes
+    struct attention_inferred
+    {
+        std::size_t batch_size; // From input dim(1)
+        std::size_t
+            input_hidden_size;   // From input dim(3) - Also known as embeddng_size in literature
+        std::size_t hidden_size; // From weights or qkv_hidden_sizes, related to q,k size which must
+                                 // be equal
+        std::size_t
+            v_hidden_size; // From weights or qkv_hiddn_sizes - differs during cross attention_mode
+        std::size_t
+            sequence_length; // Pulled from input dim(1), must be consistent between other parms
+        std::size_t max_sequence_length;   // Pulled from mask_index
+        std::size_t total_sequence_length; // Pulled from past_seq_length + sequence_length
+        std::size_t query_size;            // Also known as head_size. Derived via num_heads
+        bool has_past_input =
+            false; // Set to true when we have past input. Present output required when set
+        bool has_input_bias = false; // Set to true when we have input_bias
+        bool has_attn_mask  = false; // Set to true when we have an attention mask set
+        bool has_attn_bias  = false; // Set to true when we have an attention bias input
+        bool scale_not_query_sz =
+            false; // Set when a different scale attribute from 1/sqrt(query_size)
+        mask_pad index_pad =
+            mask_pad::no_pad; // Used to track state of input projection mask padding
+        atten_mode attn_type; // Used to determine the attention configuration
+    };
+
+    static void handle_qkv_hidden_size_attr(const onnx_parser& parser,
+                                            const onnx_parser::node_info& info,
+                                            struct attention_attr& attr_out)
```

Comment:
The `struct` keyword should not be needed here. You do that several places and I think it can be removed everywhere.

Reply:
Okay will clean this up